lib/autoupdate/agent/setup.go

```diff
@@ -273,11 +289,198 @@ func (ns *Namespace) Setup(ctx context.Context, path string) error {
+	}
+
+	cmd := localExec{
+		Dir:      tempDir,
+		Log:      ns.log,
+		ErrLevel: slog.LevelDebug,
+		OutLevel: slog.LevelDebug,
+	}
+
+	// Build and install the module, then ensure files are properly labeled.
+	_, err = cmd.Run(ctx, "make", "-f", selinuxMakefile, "teleport.pp")
+	if err != nil {
+		return trace.Wrap(err, "failed to build module")
+	}
+
+	_, err = cmd.Run(ctx, "semodule", "-i", "teleport.pp")
+	if err != nil {
+		return trace.Wrap(err, "failed to install module")
+	}
+
+	if err := ns.createAndLabelDirs(ctx, cmd); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (ns *Namespace) createAndLabelDirs(ctx context.Context, cmd localExec) error {
+	// Create the teleport-upgrade and data dirs ahead of time so we
+	// can ensure they exist when we label them.
+	if err := os.MkdirAll(versioncontrol.UnitConfigDir, defaults.DirectoryPermissions); err != nil {
+		ns.log.WarnContext(ctx, "Failed to create teleport-upgrade directory.", errorKey, err)
+	}
+	if err := os.MkdirAll(ns.dataDir, teleport.PrivateDirMode); err != nil {
+		ns.log.WarnContext(ctx, "Failed to create teleport data directory.", errorKey, err)
+	}
+
+	dirsToLabel := []string{
+		filepath.Clean(ns.installDir),
+		ns.dataDir,
+		versioncontrol.UnitConfigDir,
+	}
+	// Create an empty teleport.yaml config file if it doesn't exist, and
+	// only attempt to label it if it exists.
+	if libutils.FileExists(ns.configFile) {
+		dirsToLabel = append(dirsToLabel, filepath.Dir(ns.configFile))
+	} else {
+		confFile, err := os.OpenFile(ns.configFile, os.O_RDONLY|os.O_CREATE, 0o666)
+		if err != nil {
+			ns.log.WarnContext(ctx, "Failed to create teleport.yaml.", errorKey, err)
+			ns.log.WarnContext(ctx, "You will likely need to create teleport.yaml and re-run 'teleport-update enable' for Teleport SSH to work correctly with SELinux.")
+		} else {
+			confFile.Close()
+			dirsToLabel = append(dirsToLabel, filepath.Dir(ns.configFile))
+			ns.log.WarnContext(ctx, "Created an empty teleport.yaml.", "path", ns.configFile)
+			ns.log.WarnContext(ctx, "If you move or copy another file onto this file you will likely need to re-run 'teleport-update enable' for Teleport SSH to work correctly with SELinux.")
+		}
+	}
+
+	labelArgs := append([]string{"-rv"}, dirsToLabel...)
+	_, err := cmd.Run(ctx, "restorecon", labelArgs...)
+	if err != nil {
+		return trace.Wrap(err, "failed to restore file contexts")
+	}
+
+	return nil
+}
+
+func (ns *Namespace) removeSELinux(ctx context.Context) error {
+	ns.log.InfoContext(ctx, "Removing SELinux module.")
+
+	if err := ns.checkSELinux(ctx, false); err != nil {
+		return trace.Wrap(err)
+	}
+
+	cmd := localExec{
+		Log:      ns.log,
+		ErrLevel: slog.LevelDebug,
+		OutLevel: slog.LevelDebug,
+	}
+
+	removeOutput, err := cmd.Output(ctx, "semodule", "-r", "teleport")
+	if err != nil {
+		// If the module is not installed, return without an error.
+		if bytes.Contains(removeOutput, []byte("(No such file or directory).")) {
```

Comment:
This seems brittle. Is there an an exist code for this case?

Reply:
Unfortunately no, all errors exit with code 1